WIP: ethminer-ProgPoW rebase - spec v0.9.2

hackmod opened the pull request:
**WARNING:** *only cuda tested.*
with https://github.com/EthersocialNetwork/go-ethereum/tree/progpow-v0.9.2
(modified version of the original work by Martin)

manually rebased on top of the current master branch.

`/share/ProgPoW/build/ethminer/ethminer --noeval -v 3 --cuda-devices 0 -U -P http://127.0.0.1:8548`
![image](https://user-images.githubusercontent.com/32324335/49693925-577ffc00-fbc3-11e8-9b5c-aca3405d70cd.png)

I'm not try to merge this PR. ***this is just a (manually) rebased code.***
but we can test / study and use it as a reference work to merge ProgPoW into the ethminer codebase.

![image](https://user-images.githubusercontent.com/32324335/49694137-4933df00-fbc7-11e8-8671-32fe1f08ba07.png)

ddobreff commented:
There are already 2 PRs for progpow
integration, can’t you just work on one?

hackmod commented:
yes. There are already ProgPoW PR by @MariusVanDerWijden and this is just a rebase work on top of the latest master branch.
this PR could be regarded as reference work

naikmyeong commented:
@ddobreff I've just closed my PR

naikmyeong commented:
@hackmod Can you fix these?

```
CMake Warning (dev) at ethminer/CMakeLists.txt:17 (find_package):
  Policy CMP0074 is not set: find_package uses <PackageName>_ROOT variables.
  Run "cmake --help-policy CMP0074" for policy details.  Use the cmake_policy
  command to set the policy and suppress this warning.
  CMake variable CLI11_ROOT is set to:
    C:/.hunter/_Base/951e8da/4717139/c15df5a/Install
  Environment variable CLI11_ROOT is set to:
    C:/.hunter/_Base/951e8da/4717139/c15df5a/Install
  For compatibility, CMake is ignoring the variable.
This warning is for project developers.  Use -Wno-dev to suppress it.
CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
Please set them or make sure they are set and tested correctly in the CMake files:
CUDA_nvrtc_LIBRARY
    linked by target "ethash-cuda" in directory C:/projects/ethminer/libethash-cuda
-- Configuring incomplete, errors occurred!
See also "C:/projects/ethminer/build/CMakeFiles/CMakeOutput.log".
See also "C:/projects/ethminer/build/CMakeFiles/CMakeError.log".
```

hackmod commented:
rebased. fixed travis error

hackmod commented:
binaries available at https://github.com/hackmod/ethminer/releases/tag/v0.9.2-progpow

- ***only CUDA supported***
- you can test it locally with `--noevel` option like as follow
`ethminer --noeval -v 3 --cuda-devices 0 -U -P http://127.0.0.1:8545`

See also progpow testnet:
 - netstats - http://boot.gangnam.ethdevops.io/
 - https://gist.github.com/holiman/c8944c8be26f798e391822b9a70cb91c#gistcomment-2783941

MariusVanDerWijden commented:
Thank you for all the work! 
However mining currently fails with 
```
Unexpected error CUDA error in func compileKernel at line 396 calling cuModuleLoadDataEx(&m_module, ptx, 6, jitOpt, jitOptVal) failed with error CUDA_ERROR_INVALID_PTX on CUDA device 65:00.0
```

hackmod commented:
> Thank you for all the work!
> However mining currently fails with
> 
> ~~~
> Unexpected error CUDA error in func compileKernel at line 396 calling cuModuleLoadDataEx(&m_module, ptx, 6, jitOpt, jitOptVal) failed with error CUDA_ERROR_INVALID_PTX on CUDA device 65:00.0
> ~~~

thankyou for your feedback~

did you download binary from my repo?

for some cases, cmake detect wrong `libnvrtc*so` path and link different version of the nvrtc and it makes some trouble with `cuModuleLoadDataEx()`

hackmod commented:
now OpenCL supported!
- [x] extract `compileKernel()`
- [x] compileKernel every 50 blocks

https://github.com/hackmod/ethminer/releases/tag/v0.9.2-opencl-win-1218
- binaries snapshot for testing: temporary disabled CUDA, windows binaries only.

hackmod commented:
fixed epoch change issue
https://github.com/hackmod/ethminer/releases/tag/v0.9.2-opencl-win-1219
- OpenCL: fixed setArg issue after epoch change
- CUDA: fixed manual rebase mistake. (fix `incompatible driver context` error after epoch change)
